Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+|---|---|---|---|---|
+| NSString, C#, Rust, Go | `"Cafe"` | `"´"` | n/a | n/a |
+| Java, Ruby, Perl | `"Cafe"` | `"´"` | `"Café"` | `""` |
+
+By and large, other languages do not use canonical equivalence when matching characters in regular expressions. Java provides a `CANON_EQ` option when compiling a pattern to opt into such checking.
+
+```java
+String cafe = "Cafe\u0301";
+
+Pattern exactPatt = Pattern.compile("é");
+exactPatt.matcher(cafe).find()      // false
+
+Pattern equivPatt = Pattern.compile("é", Pattern.CANON_EQ);
+equivPatt.matcher(cafe).find()      // true
+```
+
+Finally, there is wide variance of the actual definitions of character classes between different languages. As one example, should the `\d` character class, which matches digits, include the digits from the _Halfwidth and Fullwidth Forms_ Unicode block? ICU-based regular expression engines, like those used in Java and Foundation's `NSRegularExpression`, include these extra digits, while other languages do not. Determining exactly what characters are matched by each character class of a particular language or library can be difficult, making it harder to fully understand what kinds of strings a pattern will or will not match.
+
+## Proposed Solution
+
+We propose a treatment of character classes in Swift regular expression that is consistent with the Swift string model, configurable at the point of declaration, and comprehensible for users getting to know Swift regular expressions.
+
+### Consistency
+
+Just as the default view of a string's contents are as a collection of grapheme-cluster characters, with comparisons using canonical equivalence, regular expression matching by default will do the same. This assures that similar code produces predictable results as we add APIs that accept regular expressions and other patterns:
+
+```swift
+let str = "Cafe\u{301}"
+str.contains("é")       // true
+str.contains(/é/)       // true
+```
+
+Moreover, in accordance with the [Unicode guidelines for regular expressions](https://www.unicode.org/reports/tr18), we use Unicode property data rather than specific lists of characters whenever possible.
+
+
+### Configuration
+
+For compatibility with other languages, and in some cases to allow users to opt into higher performance, you can configure the character class matching level of regular expression at the declaration site. For example, you can opt into the Unicode scalar-based matching of other languages by using the `unicodeScalarSemantics` property on a regular expression instance.
+
+This example declares `pattern` with Unicode scalar semantics, instead of the default character semantics. `pattern` doesn't match the string above because `"e"` and `"\u{301}"` are treated as separate values.
+
+```swift
+let pattern = /^Caf.$/.unicodeScalarSemantics
+str.contains(pattern)   // false
+```
+
+Initially, we plan to allow configuration between character, Unicode scalar, and POSIX-compatible character class matching.
+
```

Comment:
Is POSIX-or-not the only character class switching option supported by those other engines?